### PR TITLE
Enable detecting Red Hat distors in PlatformDetection

### LINKED ISSUE #21920: Test: System.IO.Tests.DirectoryInfo_Delete_bool/Unix_NotFoundDirectory_ReadOnlyVolume failed with "Xunit.Sdk.TrueException"
Opened on behalf of @Jiayili1

The test `System.IO.Tests.DirectoryInfo_Delete_bool/Unix_NotFoundDirectory_ReadOnlyVolume` has failed.

Assert.True() Failure
    Expected: True
    Actual:   False

        Stack Trace:

           at System.IO.Tests.FileSystemTest.RunAsSudo(String commandLine) in /root/corefx/src/System.IO.FileSystem/tests/FileSystemTest.cs:line 69
       at System.IO.Tests.FileSystemTest.ReadOnly_FileSystemHelper(Action`1 testAction, String subDirectoryName) in /root/corefx/src/System.IO.FileSystem/tests/FileSystemTest.cs:line 98
       at System.IO.Tests.Directory_Delete_str.Unix_NotFoundDirectory_ReadOnlyVolume() in /root/corefx/src/System.IO.FileSystem/tests/Directory/Delete.cs:line 214
Build : Master - 20170706.01 (Core Tests)
Failing configurations:
- Ubuntu.1404.Amd64-x64
  - Debug

Detail: https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20170706.01/workItem/System.IO.FileSystem.Tests/analysis/xunit/System.IO.Tests.DirectoryInfo_Delete_bool~2FUnix_NotFoundDirectory_ReadOnlyVolume

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #21920 and #23630

### CODE REVIEW

**Comment by davidsh** [comment]
>NETFX CI failure: System.Diagnostics.StackTrace

>-2a8f9c29\src\System.Diagnostics.StackTrace\tests\System.Diagnostics.StackTrace.Tests.csproj]
10:43:22 D:\j\workspace\windows-TGrou---2a8f9c29\Tools\tests.targets(483,5): error : One or more tests failed while running tests from 'System.Diagnostics.StackTrace.Tests' please check D:\j\workspace\windows-TGrou---2a8f9c29\bin/AnyOS.AnyCPU.Release/System.Diagnostics.StackTrace.Tests/netfx/testResults.xml for details! [D:\j\workspace\windows-TGrou---2a8f9c29\src\System.Diagnostics.StackTrace\tests\System.Diagnostics.StackTrace.Tests.csproj]
10:43:22 Command execution failed with exit code 1.

**Comment by tarekgh** [comment]
test NETFX x86 Release Build please

**Comment by tarekgh** [comment]
The failures in NETFX x86 Release Build is unrelated and tracked by the issue

https://github.com/dotnet/corefx/issues/23796

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs [hunk 1]
```diff
@@ -103,6 +103,77 @@ private static DistroInfo ParseOsReleaseFile()
...
+                                i++;
+
+                            if (i < line.Length)
+                            {
+                                int j = i + 1;
+                                while (j < line.Length && (Char.IsNumber(line[j]) || line[j] == '.'))
+                                    j++;
+
+                                ret.VersionId = line.Substring(i, j - i);
+                                ret.Version = line.Substring(i, line.Length - i);
```
Comment (danmosemsft) [inline src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: `.Trim()` ?

Comment (tarekgh) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: ret.Version is not really the real version but something telling about the distro sku (something like "7.3 (Maipo)"). we are not going to use it officially for getting the version. instead, we use ret.VersionId. so trimming is not really necessary.

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs [hunk 2]
```diff
@@ -103,6 +103,77 @@ private static DistroInfo ParseOsReleaseFile()
...
+                    using (StreamReader file = new StreamReader(fileName))
+                    {
+                        string line = file.ReadLine();
+                        if (!String.IsNullOrEmpty(line))
+                        {
+                            if (line.StartsWith("Red Hat Enterprise Linux", StringComparison.OrdinalIgnoreCase))
+                            {
+                                ret.Id = "rhel";
+                            }
+                            else if (line.StartsWith("Centos", StringComparison.OrdinalIgnoreCase))
```
Comment (danmosemsft) [inline src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: Centos seems to successfully work using /etc/os-release, at least at https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20170905.03/workItem/System.Runtime.InteropServices.RuntimeInformation.Tests/wilogs I see 
`2017-09-05 16:38:03,416: INFO: proc(54): run_and_log_output: Output: Distro=centos VersionId=7 Pretty=CentOS Linux 7 (Core) Version=7 (Core) OS=Linux 3.10.0-514.26.2.el7.x86_64 #1 SMP Tue Jul 4 15:04:05 UTC 2017 OSVer=Unix 3.10.0.514 OSArch=X64 Arch=X64 Framework=.NET Core 4.6.25701.02`

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs [hunk 3]
```diff
@@ -103,6 +103,77 @@ private static DistroInfo ParseOsReleaseFile()
                     }
                 }
             }
+            else
```
Comment (danmosemsft) [inline src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: can you also add 
```
PlatformDetection.IsRHEL
PlatformDetection.IsNotRHEL
PlatformDetection.IsRHEL9
PlatformDetection.IsNotRHEL9
```
or whatever you think we'll need?

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs (hunk 1 above)
Comment (danmosemsft) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: Right, just tidy...

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs (hunk 2 above)
Comment (tarekgh) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: Yes, Centos currently work. but Centos is the future can decide to do the same as Red Hat. especially it is built from the same sources. currently, in Centos, it contains the same file we read the version info as the one in Red Hat. the current change will not affect the current behavior as we'll try to read the version on Centos the same way we do today. The change here is just a fallback path.

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs (hunk 3 above)
Comment (tarekgh) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: we'll add these when needed. do you know any place today need it? just curios.

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs (hunk 2 above)
Comment (danmosemsft) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: sounds good

File: src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs (hunk 3 above)
Comment (tarekgh) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: ok, I'll add these according to your comment.

Comment (danmosemsft) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: https://github.com/dotnet/corefx/issues/21920 perhaps ? 
https://github.com/dotnet/corefx/issues/23630
you could fix both of those if you like

for the globalization ones, I guess you're using ICU version, which is better.

Comment (danmosemsft) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: just one; https://github.com/dotnet/corefx/issues/23628

Comment (tarekgh) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: The Globalization one I'll fix it with checking ICU version as this is more reliable. 

For the other 2 issues, do we just need to skip these tests on red hat? or we need to do something more? 

CC @JeremyKuhne for System.IO issues

Comment (danmosemsft) [reply src/CoreFx.Private.TestUtilities/src/System/PlatformDetection.Unix.cs]: I think #23630 does as `CanGetSetPipeSz` is returning false. I don't know about #21920 - maybe RHEL doesn't support read only volumes. @ianhays

**Comment by danmosemsft** [comment]
@dotnet-bot test linux x64 release build please (https://github.com/dotnet/corefx/issues/23811)

**Pull request merged**